### PR TITLE
fix: split large Content-Security-Policy headers over multiple HTTP headers

### PR DESCRIPTION
**Pull request by n-iv**
## Summary

When the `Content-Security-Policy` header exceeds 7800 bytes, split it into multiple `header()` calls. Apache `mod_proxy_fcgi` otherwise fails with `AH01070 Error parsing script headers` and returns HTTP 500 because
`HUGE_STRING_LEN` is hardcoded at 8192 bytes in `httpd.h` and is used to parse FCGI response headers. Raising this limit by recompiling Apache does not fix the issue (other related FCGI buffer limits sit at the same value).

The split is performed only between directives, never inside one. Per CSP Level 3 §3.1, multiple CSP headers are enforced as the intersection of the conveyed policies, so behavior is unchanged for compliant clients.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)
- [ ] [Labels added](https://github.com/nextcloud/server/labels) where applicable (ex: bug/enhancement, `3. to review`, feature component)
- [ ] [Milestone added](https://github.com/nextcloud/server/milestones) for target branch/version (ex: 32.x for `stable32`)

## AI (if applicable)

- [x] The content of this PR was partly or fully generated using AI

### CODE REVIEW

File: lib/private/AppFramework/Http/Output.php [hunk 1]
```diff
@@ -46,7 +46,32 @@ public function setReadfile($path) {
...
+												   header($prefix . ' ' . $segment . ';', $first);
+												   $first = false;
+												   $segment = $directive;
+										   } else {
+												   $segment = $candidate;
+										   }
+								   }
+								   if ($segment !== '') {
+										   header($prefix . ' ' . $segment . ';', $first);
+								   }
```
[the remark this answers is not in the document]
Comment (n-iv) [reply lib/private/AppFramework/Http/Output.php]: You would need 100+ federated servers with about 25 chars each to reach that.

File: lib/private/AppFramework/Http/Output.php [hunk 2]
```diff
@@ -46,7 +46,32 @@ public function setReadfile($path) {
 	 */
 	#[\Override]
 	public function setHeader($header) {
-		header($header);
+		   $maxLen = 7800;
+		   if (strlen($header) > $maxLen) {
+				   foreach (['Content-Security-Policy:', 'Feature-Policy:'] as $prefix) {
```
[the remark this answers is not in the document]
Comment (n-iv) [reply lib/private/AppFramework/Http/Output.php]: Added a comment in next commit

**Comment by kesselb** [comment]
cc @skjnldsv could be related to https://github.com/nextcloud/documentation/pull/13803

**Comment by skjnldsv** [comment]
> cc @skjnldsv could be related to https://github.com/nextcloud/documentation/pull/13803

Was thinking the same!!

File: lib/private/AppFramework/Http/Output.php [hunk 3]
```diff
@@ -46,6 +46,34 @@ public function setReadfile($path) {
 	 */
 	#[\Override]
 	public function setHeader($header) {
+		// Apache mod_proxy_fcgi parses FCGI response headers with a buffer
+		// hardcoded at HUGE_STRING_LEN (8192 bytes in httpd.h); 7800 leaves
+		// a safety margin for the status line and surrounding header bytes.
+		$maxLen = 7800;
+		if (strlen($header) > $maxLen) {
+			foreach (['Content-Security-Policy:', 'Feature-Policy:'] as $prefix) {
+				if (strncmp($header, $prefix, strlen($prefix)) === 0) {
```
Comment (provokateurin) [inline lib/private/AppFramework/Http/Output.php]:
```suggestion
				if (str_starts_with($header, $prefix)) {
```

File: lib/private/AppFramework/Http/Output.php [hunk 4]
```diff
@@ -46,6 +46,34 @@ public function setReadfile($path) {
...
+					$value = ltrim(substr($header, strlen($prefix)));
+					$directives = array_filter(array_map('trim', explode(';', $value)));
```
Comment (provokateurin) [inline lib/private/AppFramework/Http/Output.php]:
```suggestion
					$directives = array_filter(array_map(trim(...), explode(';', $value)));
```

File: lib/private/AppFramework/Http/Output.php [hunk 5]
```diff
@@ -46,6 +46,34 @@ public function setReadfile($path) {
...
+							header($prefix . ' ' . $segment . ';', $first);
+							$first = false;
+							$segment = $directive;
+						} else {
+							$segment = $candidate;
+						}
+					}
+					if ($segment !== '') {
+						header($prefix . ' ' . $segment . ';', $first);
+					}
```
Comment (provokateurin) [inline lib/private/AppFramework/Http/Output.php:73]: This logic is way too messy and I'm pretty sure it's also not doing its job correctly.

File: lib/private/AppFramework/Http/Output.php [hunk 6]
```diff
@@ -46,6 +46,34 @@ public function setReadfile($path) {
 	 */
 	#[\Override]
 	public function setHeader($header) {
+		// Apache mod_proxy_fcgi parses FCGI response headers with a buffer
+		// hardcoded at HUGE_STRING_LEN (8192 bytes in httpd.h); 7800 leaves
+		// a safety margin for the status line and surrounding header bytes.
+		$maxLen = 7800;
```
Comment (provokateurin) [inline lib/private/AppFramework/Http/Output.php:53]: I really don't like this magic value, it should very likely just be 8192 or 8191.

**Review by provokateurin — CHANGES_REQUESTED** [review]

File: lib/private/AppFramework/Http/Output.php (hunk 5 above)
Comment (n-iv) [reply lib/private/AppFramework/Http/Output.php:73]: The splitter has been running in production for quite some time on a NC32 instance with ~23 trusted servers (CSP ~6.3 kB, splits into two policies), with no observed regression on federated shares, iframes, or form submissions. The output of the production curl is available if useful.

File: lib/private/AppFramework/Http/Output.php (hunk 6 above)
Comment (n-iv) [reply lib/private/AppFramework/Http/Output.php:53]: 7800 is indeed empirical. Just using 8192 isn't quite right though: HUGE_STRING_LEN is the buffer Apache reserves for the full header line (name + ": " + value + CRLF), so the value alone has to fit in roughly 8192 - 27 = 8165 bytes to stay safely under the parser's threshold.

We can express the intent in the code with:
private const HEADER_LINE_MAX = 8192;  // Apache HUGE_STRING_LEN
private const HEADER_NAME_OVERHEAD = 32;  // name + ": " + CRLF + margin

Or we could keep a single constant CSP_VALUE_MAX = 8000 (or any round number below 8165) with a comment pointing at HUGE_STRING_LEN

**Comment by skjnldsv** [comment]
@n-iv could you squash your commits into one again please? :)
I added some more coworkers to see if they can review. Please note that your PR is a good idea I think, but we're right in the middle of RC/final releases, so our availability is a bit scarce a the moment :wink:

**Comment by n-iv** [comment]
Hi @skjnldsv 

I squashed my commits. Np for your availability, mine is quite low too :)

**Comment by skjnldsv** [comment]
Soooo, I had another look:

The split logic only breaks between directives, never inside one. Makes sense, if you split one directive's list of allowed sources across two headers, the browser treats each header as a seperate policy and only allows what both agree on, so splitting mid directive would probably end up blocking stuff that shoud be allowed. But that also means if a single directive (`script-src`, `connect-src`, `img-src` etc, these get built up from every app's allowed domains) is itself over ~7800 bytes on its own, this fix doesnt help. It'll just emit that one oversized directive as is and were back to the same AH01070 error. Given how many apps push domains onto those directives (federation, external storage, collabora, s3 previews...) I think thats a pretty plausible real world case, not just a rare edge case. Could you test with a policy where one single directive alone is already past the limit, rather then lots of small ones adding up togehter?

Also, are the domain lists deduped anywhere before `buildPolicy() `runs? If two apps add the same domain it'll show up twice and waste bytes for free, might buy some headroom cheap.

Last thing, tests are unchecked in the PR checklist and `OutputTest.php` currently has zero coverage of `setHeader() `at all. This will need tests before merge imho, normal header passthrough, the multi header split case, and ideally the one giant directive case too, however that ends up handled.
